src/solution.cpp

```diff
@@ -0,0 +1,176 @@
+    if(node == nullptr) return 0;
+    return node->height;
+}
+int getBalance(Node* node){
+    if(node == nullptr) return 0;
+    return getHeight(node->right) - getHeight(node->left);
+}
+
+Node* rotateLeft(Node* node){
+    Node* tempR = node->right;
+    Node* temp2 = tempR->left;
+
+    tempR->left = node;
+    node->right = temp2;
+
+    node->height = 1 + max(getHeight(node->left), getHeight(node->right));
+    tempR->height = 1 + max(getHeight(tempR->left), getHeight(tempR->right));
+
+    return tempR;
+}
+Node* rotateRight(Node* node){
+    Node* tempL = node->left;
+    Node* temp2 = tempL->right;
+
+    tempL->right = node;
+    node->left = temp2;
+
+    node->height = 1 + max(getHeight(node->left), getHeight(node->right));
+    tempL->height = 1 + max(getHeight(tempL->left), getHeight(tempL->right));
+
+    return tempL;
+}
+
+Node* maxValueNode(Node* node){
+    if(node->right == nullptr) return node;
+    else return maxValueNode(node->right);
+}
+
+Node* rebalanceAVL(Node* node, int value){
+    int balance = getBalance(node);
+
+    if(balance < -1 && value < node->left->data) return rotateRight(node);
+
+    if(balance > 1 && value > node->right->data) return rotateLeft(node);
+
+    if(balance < -1 && value > node->left->data){
+        node->left = rotateLeft(node->left);
+        return rotateRight(node);
+    }
+    if(balance > 1 && value < node->right->data){
+        node->right = rotateRight(node->right);
+        return rotateLeft(node);
+    }
+
+    return node;
+}
+
+Node* insertAVL(Node* node, int value){
+    if(node == nullptr) return new Node(value);
+    else if(value < node->data) node->left = insertAVL(node->left, value);
+    else node->right = insertAVL(node->right, value);
+
+    node->height = 1 + max(getHeight(node->left), getHeight(node->right));
+    return rebalanceAVL(node, value);
+}
+
+Node* removeAVL(Node* node, int value){
+    if(node == nullptr) return node;
+    if(value < node->data) removeAVL(node->left, value);
+    else if(value > node->data) removeAVL(node->right, value);
+    else{
+        if(node->left == nullptr || node->right == nullptr){
+            Node* temp;
+            if(node->left == nullptr) temp = node->right;
+            else temp = node->left;
+
+            if(temp == nullptr){ // case when both left and right are null
+                temp = node;
+                node = nullptr;
+            } else node->data = temp->data; // case when only one is null and other is not
+            delete temp;
+        } else { // case when both are not null
+            // here temp is the highest number on the left, but I could also use the lowest on the right
+            Node* temp = maxValueNode(node->left);
+            node->data = temp->data;
+            node->left = removeAVL(node->left, temp->data);
+        }
+    }
+    if(node == nullptr) return node;
+    node->height = 1 + max(getHeight(node->left), getHeight(node->right));
+    return rebalanceAVL(node, value);
```

Comment:
Please, re-balancing should be before calculating of height

Reply:
hm, that makes sense, im gonna fix it in a minute
btw i wanted to ask something about this part of the code, i wrote the rebalanceAVL() function for the insert and i think it does its job there and its just a small fix to put it before re-calculating the height, but i am also using the rebalanceAVL function in the remove and as the value i just put the value of the node i deleted, could you tell me if the remove is okay like that or should i write another function for rebalancing in the remove, thanks in advance